### DISCUSSION TITLE
Behaviours

### DISCUSSION THREAD

**Discussion by jamesfoster** [nunit/nunit#3800] [Show and tell]
Hi,

I've just released a new pet project based on NUnit. It's in the very early stages and I'd like some feedback and assistance if anyone can.

The idea is to introduce the "Given, When, Then" style of writing tests to NUnit in a simple syntax.

https://github.com/jamesfoster/NUnit.Behaviours

I create a new fixture for each behaviour and one test per assertion. I generate the name of the tests from the expressions used in the steps. When executing the tests, I only run the setup once for a behaviour (assertions should not change the state). This gives clean, composable and efficient setup and and nice declarative style of testing.

I've put a screenshot of the current output of an example test in the readme. You can see the test names are very long and in practice could be much longer with more complicated setups. I'd like to nest the `given` and `when` steps. Can anyone point me to an example of how to create a nested structure of tests.

Thanks.

Comments/criticism welcome.

**Comment by jamesfoster** [comment]
Any thoughts on this? @rprouse ?

**Reply by rprouse** [reply]
This looks very interesting project. I like the idea.

To get the behaviour you want, a test fixture should contain child test fixtures for the given. If you look at any test method in NUnit that generates multiple tests like multiple `TestCase` attributes on a method, a `TestFixture` is created and then each test case becomes a test under that.
